runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPortablePipelineTranslator.java

```diff
@@ -423,8 +432,133 @@ private void translateImpulse(
           String id,
           RunnerApi.Pipeline pipeline,
           StreamingTranslationContext context) {
+    // TODO: Fail on stateful DoFns for now.
+    // TODO: Support stateful DoFns by inserting group-by-keys where necessary.
+    // TODO: Fail on splittable DoFns.
+    // TODO: Special-case single outputs to avoid multiplexing PCollections.
+    RunnerApi.Components components = pipeline.getComponents();
+    RunnerApi.PTransform transform = components.getTransformsOrThrow(id);
+    Map<String, String> outputs = transform.getOutputsMap();
+    RehydratedComponents rehydratedComponents =
+            RehydratedComponents.forComponents(components);
+
+    BiMap<String, Integer> outputMap =
+            FlinkPipelineTranslatorUtils.createOutputMap(outputs.keySet());
+    Map<String, Coder<WindowedValue<?>>> outputCoders = Maps.newHashMap();
+    for (String localOutputName : new TreeMap<>(outputMap.inverse()).values()) {
+      String collectionId = outputs.get(localOutputName);
+      Coder<WindowedValue<?>> windowCoder = (Coder) instantiateCoder(collectionId, components);
+      outputCoders.put(localOutputName, windowCoder);
+    }
+
+    final RunnerApi.ExecutableStagePayload stagePayload;
+    try {
+      stagePayload = RunnerApi.ExecutableStagePayload.parseFrom(transform.getSpec().getPayload());
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+
+    String inputPCollectionId =
+            Iterables.getOnlyElement(transform.getInputsMap().values());
```

Comment:
Unfortunately the way in which we construct coders and other properties of the ExecutableProcessBundleDescriptor are done during execution and it would be best if we somehow could make all these details be stable during pipeline translation so during execution it doesn't change. Having Flink rely on calling WireCoders.instantiateRunnerWireCoder(...) is an anti-pattern for encapsulation.

Reply:
Unfortunately, Flink needs to have an associated serializer (TypeInformation, aka Coder) with each distributed collection. This TypeInformation needs to be known at pipeline construction time. It need not match the exact coder being used to materialize elements over gRPC, but it does need to match the in-memory element type.

We could get around this partially by representing everything as bytes. The downside is that each runner-native operation that requires structure (e.g., GBK) will require an additional operation to break elements into their constituent parts. This step itself also requires knowledge of the coded type, so we ultimately run into the same issue.

Reply:
I'm just arguing for pushing most of the manipulation done within ExecutableProcessBundleDescriptor into the ExecutableStage payload (minus the ApiServiceDescriptor binding) so it doesn't need modification. This would allow the ExecutableStage to concretely answer what are the input coders, output coders, side input coders, state coders, ... in addition to any other information.

Longer term it seems if we had a way for the runner to say whether we need a keyed input context or grouped keyed output context makes sense as the runner could then say. These are the cases I know of:
* KV<Key, Value> for SplittableDoFn input, StatefulDoFn input, GBK input, Multimap side input materialization input, window mapping input
* KV<Key, Iterable<Value>> for GBK output
Do you know of any others?

Reply:
I added a test for serialization. If we agree on the repeated instantiation of ExecutableStage, then I can take this up in a separate PR (for both, batch and streaming translation). I would do that once we have test and end-to-end coverage, right now the translators are still not wired.

Reply:
SGTM